### PR TITLE
AlertingNG: unified alerting rules list

### PR DESCRIPTION
**Pull request by domasx2**
**What this PR does / why we need it**:
Work in progress on the unified alerting rules list page.
- [x] load rules from multiple rules sources
- [x] render namespaces, groups and rules
- [x] tests

list yay
Screenshot at 2021-03-22 15-45-40

**Which issue(s) this PR fixes**:
https://github.com/grafana/alerting-squad/issues/19

### CODE REVIEW

File: public/app/features/alerting/unified/state/actions.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+import { DataSourceType } from '../utils/datasource';
+
+/*
+ * Will need to be updated to:
+ *
+ * 1. Fetch grafana managed rules when the endpoint becomes available
+ * 2. Reconcile with rules from the ruler where ruler is available
+ */
+
+export const fetchRulesAction = createAsyncThunk(
```
Comment (domasx2) [inline public/app/features/alerting/unified/state/actions.ts:14]: so this was poorly thought out by me. If request for one of the rules sources is slow/hangs, user has to wait for it to finish before being able to view any of the rules. 
Need to load rules for each rules source separately, back to the drawing board

Comment (domasx2) [reply public/app/features/alerting/unified/state/actions.ts:13]: done

File: public/app/features/alerting/unified/components/TimeToNow.tsx [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import React, { FC, useEffect, useState } from 'react';
+
+export interface Props {
+  date: DateTimeInput;
+}
+
+export const TimeToNow: FC<Props> = ({ date }) => {
+  const setRandom = useState(0)[1];
+  useEffect(() => {
+    const interval = setInterval(() => setRandom(Math.random()), 1000);
```
Comment (nathanrodman) [inline public/app/features/alerting/unified/components/TimeToNow.tsx:11]: Could you explain what this effect is doing?

**Review by nathanrodman — APPROVED** [review]
It looks good to me. It would be nice if we could also include some tests but I understand this is already a very large PR behind a feature toggle so I'm fine if we commit to adding those later to unblock further development.

File: public/app/features/alerting/unified/components/TimeToNow.tsx (hunk 2 above)
Comment (domasx2) [reply public/app/features/alerting/unified/components/TimeToNow.tsx:11]: it's setting state every second to force a re-render and update the date readout

**Comment by domasx2** [comment]
> It looks good to me. It would be nice if we could also include some tests but I understand this is already a very large PR behind a feature toggle so I'm fine if we commit to adding those later to unblock further development.

I am working on tests right now! :factory_worker:

File: public/app/features/alerting/unified/components/StateTag.tsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import { useStyles } from '@grafana/ui';
+import { PromAlertingRuleState } from 'app/types/unified-alerting/dto';
+import { css, cx } from 'emotion';
+import React, { FC } from 'react';
+
+type Props = {
+  status: PromAlertingRuleState;
+};
+
+const getStyles = (theme: GrafanaTheme) => ({
```
Comment (peterholmberg) [inline public/app/features/alerting/unified/components/StateTag.tsx:11]: I'd like to move the styles under the component as in all other components.

File: public/app/features/alerting/unified/components/StateColoredText.tsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import { useStyles } from '@grafana/ui';
+import { PromAlertingRuleState } from 'app/types/unified-alerting/dto';
+import { css } from 'emotion';
+import React, { FC } from 'react';
+
+type Props = {
+  status: PromAlertingRuleState;
+};
+
+const getStyles = (theme: GrafanaTheme) => ({
```
Comment (peterholmberg) [inline public/app/features/alerting/unified/components/StateColoredText.tsx:11]: I'd like to move the styles under the component as in all other components.

**Review by peterholmberg — CHANGES_REQUESTED** [review]
Ok, this took a while. Substantial work but let's not do PRs like this again 😅 

Left a few comments. Another thing that I'd like to mention is that data source is two words and should be spelt `DataSource` or `dataSource` in the code.

File: public/app/features/alerting/unified/components/AlertLabel.tsx [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+export const AlertLabel: FC<Props> = ({ labelKey, value }) => (
+  <div className={useStyles(getStyles)}>
+    {labelKey}={value}
+  </div>
+);
+
+export const getStyles = (theme: GrafanaTheme) => css`
+  padding: 3px ${theme.spacing.xs};
```
[the remark this answers is not in the document]
Comment (domasx2) [reply public/app/features/alerting/unified/components/AlertLabel.tsx:18]: Figma does not render html + css, it draws on a `<canvas />`, hah!

It looks normal to me, hence I didn't notice: 
image

But what I **think** is happening: 
* There are no glyphs for font weight 600 imported for Roboto: https://github.com/grafana/grafana/blob/master/public/sass/base/_fonts.scss 
* Theme defines `bold` as `600`: https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/themes/default.ts#L56
* Browser attempts to interpolate to 600 from another weight, and in some cases (browser? OS? resolution?) this produces blurry results


Could you verify that it looks ok for you with semibold (500) and open an issue in grafana repo with details of your environment?

File: public/app/features/alerting/unified/components/rules/ActionButton.tsx [hunk 6]
```diff
@@ -0,0 +1,16 @@
+import { Button, ButtonProps } from '@grafana/ui/src/components/Button';
```
[the remark this answers is not in the document]
Comment (domasx2) [reply public/app/features/alerting/unified/components/rules/ActionButton.tsx:1]: It apply not only styles but also some default props and is reused in several places. 
Do you think just the styles should be applied, and props repeated?

File: public/app/features/alerting/unified/utils/formatting.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+function limitFloatingPrecision(x: number, precision: number): string {
+  return parseFloat(x.toFixed(precision)).toString();
+}
+
+/**
+ * Formats a duration provided in seconds. If smaller than 0, the most suitable
+ * unit will be picked. Truncated at precision, not rounded.
+ * (Ex.: (0.043678, 2) => '43.67 ms')
+ */
+export function formatDuration(duration: number, precision = 1): string {
```
[the remark this answers is not in the document]
Comment (domasx2) [reply public/app/features/alerting/unified/utils/formatting.ts]: wasn't aware of it! Updated top use `rangeUtils.secondsToHms`. It doesn't have microseconds, so lots of "less than a millisecond" for trivial rules. But probably ok, somebody will only care if it's slow

File: public/app/features/alerting/unified/components/StateTag.tsx (hunk 3 above)
Comment (domasx2) [reply public/app/features/alerting/unified/components/StateTag.tsx:17]: done!

File: public/app/features/alerting/unified/components/StateColoredText.tsx (hunk 4 above)
Comment (domasx2) [reply public/app/features/alerting/unified/components/StateColoredText.tsx:17]: done!

File: public/app/features/alerting/unified/components/rules/AlertInstancesTable.tsx [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+                <td>
+                  <CollapseToggle isCollapsed={!isExpanded} onToggle={() => toggleExpandedState(key)} />
+                </td>
+                <td>
+                  <StateTag status={instance.state} />
+                </td>
+                <td className={styles.labelsCell}>
+                  <AlertLabels labels={instance.labels} />
+                </td>
+                <td className={styles.createdCell}>{String(instance.activeAt).substr(0, 19).replace('T', ' ')}</td>
```
[the remark this answers is not in the document]
Comment (domasx2) [reply public/app/features/alerting/unified/components/rules/AlertInstancesTable.tsx]: shameful defensive programming :) removed

File: public/app/features/alerting/unified/components/rules/ActionIcon.tsx [hunk 9]
```diff
@@ -0,0 +1,31 @@
+import { Icon, IconName, useStyles } from '@grafana/ui';
+import { Tooltip, TooltipProps } from '@grafana/ui/src/components/Tooltip/Tooltip';
```
[the remark this answers is not in the document]
Comment (domasx2) [reply public/app/features/alerting/unified/components/rules/ActionIcon.tsx]: done

File: public/app/features/alerting/unified/components/rules/ActionIcon.tsx [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+import React, { FC } from 'react';
+import { css } from 'emotion';
+
+interface Props {
+  tooltip: TooltipProps['content'];
```
[the remark this answers is not in the document]
Comment (domasx2) [reply public/app/features/alerting/unified/components/rules/ActionIcon.tsx]: done

File: public/app/features/alerting/unified/components/rules/ActionIcon.tsx [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+  icon: IconName;
+
+  tooltipPlacement?: TooltipProps['placement'];
```
[the remark this answers is not in the document]
Comment (domasx2) [reply public/app/features/alerting/unified/components/rules/ActionIcon.tsx]: done

File: public/app/features/alerting/unified/components/AlertingPageWrapper.tsx [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    <Page navModel={navModel} className={styles.page}>
+      <Page.Contents isLoading={isLoading}>{children}</Page.Contents>
+    </Page>
+  );
+};
+
+export const getStyles = () => ({
+  page: css`
+    .page-container {
+      max-width: 1400px;
```
[the remark this answers is not in the document]
Comment (domasx2) [reply public/app/features/alerting/unified/components/AlertingPageWrapper.tsx]: default width was too short for the designs, so we agreed with Oana to make it `1400px`. However I know changed it to `theme.breakpoints.xxl` which is `1440px`. I think we'll align the other pages as we progress

File: public/app/features/alerting/unified/components/AlertLabel.tsx [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+  border-radius: ${theme.border.radius.sm};
+  border: solid 1px #343b40; // @TODO
```
[the remark this answers is not in the document]
Comment (domasx2) [reply public/app/features/alerting/unified/components/AlertLabel.tsx]: done. Though it doesn't look like in the design now, theme does not have the color from Figma

**Comment by domasx2** [comment]
> Ok, this took a while. Substantial work but let's not do PRs like this again 
Yeah, I agree :) 

> Left a few comments. Another thing that I'd like to mention is that data source is two words and should be spelt `DataSource` or `dataSource` in the code.

Renamed everywhere!

**Comment by domasx2** [comment]
gonna merge this to keep moving, @peterholmberg  let's continue addressing your concerns when you're back

**Pull request merged**